library/src/cache/patch_manager.rs

```diff
@@ -244,28 +244,35 @@ impl PatchManager {
             .with_context(|| format!("Failed to delete patch dir {}", &patch_dir.display()))
     }
 
-    /// Attempts to use the last successfully booted patch as the next boot patch. If the last successfully
-    /// booted patch is not bootable or has the same number as the patch we're falling back from, we clear it.
-    fn try_fall_back_to_last_booted_patch(&mut self) {
+    /// Deletes artifacts for the provided bad_patch_number and attempts to set the next_boot_patch to the last
+    /// successfully booted patch. If the last successfully booted patch is not bootable or has the same number
+    /// as the patch we're falling back from, we clear it as well.
+    fn try_fall_back_from_patch(&mut self, bad_patch_number: usize) {
+        // No need to log failure – delete_patch_artifacts logs for us.
+        let _ = self.delete_patch_artifacts(bad_patch_number);
+
         if let Some(next_boot_patch) = self.patches_state.next_boot_patch {
-            // If we have a next_boot_patch that we're falling back from, delete its artifacts.
-            let _ = self.delete_patch_artifacts(next_boot_patch.number);
-            self.patches_state.next_boot_patch = None;
-
-            if let Some(last_boot_patch) = self.patches_state.last_booted_patch {
-                if last_boot_patch.number == next_boot_patch.number {
-                    // If the last booted patch is the same as the next boot patch, clear it.
-                    self.patches_state.last_booted_patch = None;
+            // If our next boot patch is bad_patch_number, clear it.
+            if next_boot_patch.number == bad_patch_number {
+                self.patches_state.next_boot_patch = None;
+
+                // If the last booted patch is the same as the next boot patch, clear it.
```

Comment:
This comment is wrong?  Also, why is this inside the if?  Should last_booted_patch be cleared anytime it's the same as bad_patch_number?  Also, why would we ever mark a last_booted_patch as bad?

Reply:
I think we shouldn't ever mark a last_booted_patch as bad, right?  Feels like something we also need a test for.

Reply:
This could be cleaned up a bit, doing that now.

> I think we shouldn't ever mark a last_booted_patch as bad, right? Feels like something we also need a test for.

If we have successfully booted from a patch whose size changes on disk (our proxy for file corruption), it has become bad and we don't want to try to boot from it again. We have a test for this: `clears_next_and_last_patches_if_both_fail_validation`. Adding comments and cleaning up now.

Reply:
Addressed